File: app/src/main/java/net/ivpn/client/common/prefs/ServersRepository.java
package net.ivpn.client.common.prefs;

import androidx.annotation.Nullable;

import net.ivpn.client.common.Mapper;
import net.ivpn.client.common.dagger.ApplicationScope;
import net.ivpn.client.rest.HttpClientFactory;
import net.ivpn.client.rest.IVPNApi;
import net.ivpn.client.rest.RequestListener;
import net.ivpn.client.rest.data.ServersListResponse;
import net.ivpn.client.rest.data.model.Server;
import net.ivpn.client.rest.data.model.ServerLocation;
import net.ivpn.client.rest.requests.common.Request;
import net.ivpn.client.vpn.Protocol;
import net.ivpn.client.vpn.ProtocolController;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.EnumMap;
import java.util.List;

import javax.inject.Inject;

@ApplicationScope
public class ServersRepository implements Serializable {

    private static final Logger LOGGER = LoggerFactory.getLogger(ServersRepository.class);

    private EnumMap<Protocol, EnumMap<ServerType, Server>> currentServers = new EnumMap<>(Protocol.class);
    private List<OnFavouriteServersChangedListener> onFavouritesChangedListeners;
    private List<OnServerListUpdatedListener> onServerListUpdatedListeners;
    private List<OnServerChangedListener> onServerChangedListeners;
    private Settings settings;
    private ProtocolController protocolController;
    private ServersPreference serversPreference;
    private HttpClientFactory httpClientFactory;

    private Request<ServersListResponse> request = null;

    @Inject
    public ServersRepository(Settings settings, HttpClientFactory httpClientFactory,
                             ProtocolController protocolController,
                             ServersPreference serversPreference) {
        this.settings = settings;
        this.protocolController = protocolController;
        this.serversPreference = serversPreference;
        this.httpClientFactory = httpClientFactory;

        init();
    }

    private void init() {
        onFavouritesChangedListeners = new ArrayList<>();
        onServerListUpdatedListeners = new ArrayList<>();
        onServerChangedListeners = new ArrayList<>();
        currentServers.put(Protocol.OPENVPN, new EnumMap<>(ServerType.class));
        currentServers.put(Protocol.WIREGUARD, new EnumMap<>(ServerType.class));
    }

    @Nullable
    public Server getCurrentServer(ServerType serverType) {
        Server server = getCurrentServers().get(serverType);
        if (server == null) {
            server = serversPreference.getCurrentServer(serverType);
            setCurrentServer(serverType, server);
        }

        if (server == null) {
            server = getDefaultServer(serverType);
            setCurrentServer(serverType, server);
        }

        return server;
    }

    public void setCurrentServer(ServerType serverType, Server server) {
        if (server == null) return;
        LOGGER.info("Set current server = " + server + " serverType = " + serverType);

        getCurrentServers().put(serverType, server);
        serversPreference.setCurrentServer(serverType, server);

        ServerType anotherServerType = ServerType.getAnotherType(serverType);
        Server serverMH = getCurrentServer(anotherServerType);
        if (!server.canBeUsedAsMultiHopWith(serverMH)) {
            serverMH = getDefaultServer(anotherServerType);
            getCurrentServers().put(anotherServerType, serverMH);
            serversPreference.setCurrentServer(anotherServerType, serverMH);
        }
    }

    @Nullable
    public Server getDefaultServer(ServerType serverType) {
        List<Server> servers = getServers(false);
        if (servers != null && !servers.isEmpty()) {
            Server anotherServer = serversPreference.getCurrentServer(ServerType.getAnotherType(serverType));
            for (Server server : servers) {
                if (server.canBeUsedAsMultiHopWith(anotherServer)) {
                    return server;
                }
            }
        }
        return null;
    }

    public boolean isServersListExist() {
        List<Server> servers = serversPreference.getServersList();
        return servers != null && !servers.isEmpty();
    }

    public List<Server> getServers(boolean isForced) {
        List<Server> servers = serversPreference.getServersList();
        if (isForced || servers == null) {
            //update server list online
            updateServerList(isForced);
            //update servers list offline
            tryUpdateServerListOffline();
            servers = serversPreference.getServersList();
        }
        return servers;
    }

    public List<ServerLocation> getLocations() {
        List<ServerLocation> locations = serversPreference.getServerLocations();
        if (locations == null) {
            updateLocations();
            locations = serversPreference.getServerLocations();
        }

        return locations;
    }

    private void updateLocations() {
        List<Server> servers = serversPreference.getServersList();
        List<ServerLocation> locations = new ArrayList<>();

        for (Server server : servers) {
            locations.add(new ServerLocation(
                    server.getCity(),
                    server.getCountryCode(),
                    server.getLatitude(),
                    server.getLongitude()
            ));
        }

        if (protocolController.getCurrentProtocol() == Protocol.OPENVPN) {
            serversPreference.putOpenVPNLocations(locations);
        } else {
            serversPreference.putWireGuardLocations(locations);
        }
    }

    public List<Server> getFavouritesServers() {
        return serversPreference.getFavouritesServersList();
    }

    public void addFavouritesServer(Server server) {
        LOGGER.info("addFavouritesServer server = " + server);
        serversPreference.addFavouriteServer(server);
        notifyFavouriteServerAdded(server);
    }

    public void removeFavouritesServer(Server server) {
        LOGGER.info("removeFavouritesServer server = " + server);
        serversPreference.removeFavouriteServer(server);
        notifyFavouriteServerRemoved(server);
    }

    private List<Server> getCachedServers() {
        return serversPreference.getServersList();
    }

    public Server getForbiddenServer(ServerType serverType) {
        boolean multiHop = settings.isMultiHopEnabled();
        if (!multiHop) return null;
        return serversPreference.getCurrentServer(ServerType.getAnotherType(serverType));
    }

    public void updateServerList(final boolean isForced) {
        LOGGER.info("Updating server list, isForced = " + isForced);
        request = new Request<>(settings, httpClientFactory, this, Request.Duration.SHORT);
        request.start(IVPNApi::getServers, new RequestListener<ServersListResponse>() {
            @Override
            public void onSuccess(ServersListResponse response) {
                LOGGER.info("Updating server list, state = SUCCESS_STR");
                LOGGER.info(response.toString());
                response.markServerTypes();
                setServerList(response.getOpenVpnServerList(), response.getWireGuardServerList());
                settings.setAntiTrackerDefaultDNS(response.getConfig().getAntiTracker().getDefault().getIp());
                settings.setAntiTrackerHardcoreDNS(response.getConfig().getAntiTracker().getHardcore().getIp());
                settings.setAntiTrackerDefaultDNSMulti(response.getConfig().getAntiTracker().getDefault().getMultihopIp());
                settings.setAntiTrackerHardcoreDNSMulti(response.getConfig().getAntiTracker().getHardcore().getMultihopIp());
                settings.setIpList(Mapper.stringFromIps(response.getConfig().getApi().getIps()));

                for (OnServerListUpdatedListener listener : onServerListUpdatedListeners) {
                    listener.onSuccess(getSuitableServers(response), isForced);
                }
            }

            @Override
            public void onError(Throwable throwable) {
                LOGGER.error("Updating server list, state = ERROR", throwable);
                for (OnServerListUpdatedListener listener : onServerListUpdatedListeners) {
                    listener.onError(throwable);
                }
            }

            @Override
            public void onError(String string) {
                LOGGER.error("Updating server list, state = ERROR", string);
                for (OnServerListUpdatedListener listener : onServerListUpdatedListeners) {
                    listener.onError();
                }
            }
        });
    }

    public void fastestServerSelected() {
        serversPreference.putSettingFastestServer(true);
        for (OnServerChangedListener listener : onServerChangedListeners) {
            listener.onServerChanged();
        }
    }

    public void serverSelected(Server server, ServerType type) {
        serversPreference.putSettingFastestServer(false);
        setCurrentServer(type, server);
        for (OnServerChangedListener listener : onServerChangedListeners) {
            listener.onServerChanged();
        }
    }

    public void tryUpdateServerListOffline() {
        LOGGER.info("Trying update server list offline from cache...");
        if (getCachedServers() != null) {
            return;
        }

        ServersListResponse response = Mapper.getProtocolServers(ServersLoader.load());
        response.markServerTypes();
        settings.setAntiTrackerDefaultDNS(response.getConfig().getAntiTracker().getDefault().getIp());
        settings.setAntiTrackerHardcoreDNS(response.getConfig().getAntiTracker().getHardcore().getIp());
        settings.setAntiTrackerDefaultDNSMulti(response.getConfig().getAntiTracker().getDefault().getMultihopIp());
        settings.setAntiTrackerHardcoreDNSMulti(response.getConfig().getAntiTracker().getHardcore().getMultihopIp());
        settings.setIpList(Mapper.stringFromIps(response.getConfig().getApi().getIps()));

        setServerList(response.getOpenVpnServerList(), response.getWireGuardServerList());
    }

    public void tryUpdateIpList() {
        if (settings.getIpList() != null) {
            return;
        }

        ServersListResponse response = Mapper.getProtocolServers(ServersLoader.load());

        settings.setAntiTrackerDefaultDNS(response.getConfig().getAntiTracker().getDefault().getIp());
        settings.setAntiTrackerHardcoreDNS(response.getConfig().getAntiTracker().getHardcore().getIp());
        settings.setAntiTrackerDefaultDNSMulti(response.getConfig().getAntiTracker().getDefault().getMultihopIp());
        settings.setAntiTrackerHardcoreDNSMulti(response.getConfig().getAntiTracker().getHardcore().getMultihopIp());
        settings.setIpList(Mapper.stringFromIps(response.getConfig().getApi().getIps()));
    }

    public void tryUpdateServerLocations() {
        LOGGER.info("tryUpdateServerLocations BEFORE");
        if (serversPreference.getServerLocations() != null) {
            return;
        }
        LOGGER.info("tryUpdateServerLocations AFTER");

        ServersListResponse response = Mapper.getProtocolServers(ServersLoader.load());
        response.markServerTypes();

        List<ServerLocation> locations = new ArrayList<>();

        for (Server server : response.getOpenVpnServerList()) {
            locations.add(new ServerLocation(
                    server.getCity(),
                    server.getCountryCode(),
                    server.getLatitude(),
                    server.getLongitude()
            ));
        }
        serversPreference.putOpenVPNLocations(locations);
        locations.clear();

        for (Server server : response.getWireGuardServerList()) {
            locations.add(new ServerLocation(
                    server.getCity(),
                    server.getCountryCode(),
                    server.getLatitude(),
                    server.getLongitude()
            ));
        }
        serversPreference.putWireGuardLocations(locations);

        setServerList(response.getOpenVpnServerList(), response.getWireGuardServerList());

        updateCurrentServersWithLocation();

        currentServers.put(Protocol.OPENVPN, new EnumMap<>(ServerType.class));
        currentServers.put(Protocol.WIREGUARD, new EnumMap<>(ServerType.class));
    }

    private void updateCurrentServersWithLocation() {
        serversPreference.updateCurrentServersWithLocation();
    }

    public void setLocationList(List<ServerLocation> openVpnLocations, List<ServerLocation> wireguardLocations) {
        LOGGER.info("Putting locations, OpenVPN locations list size = " + openVpnLocations.size() + " WireGuard = " + wireguardLocations.size());
        serversPreference.putOpenVPNLocations(openVpnLocations);
        serversPreference.putWireGuardLocations(wireguardLocations);
    }

    public void setServerList(List<Server> openvpnServers, List<Server> wireguardServers) {
        LOGGER.info("Putting servers, OpenVpn servers list size = " + openvpnServers.size() + " WireGuard = " + wireguardServers.size());
        serversPreference.putOpenVpnServerList(openvpnServers);
        serversPreference.putWireGuardServerList(wireguardServers);
    }

    public void addToExcludedServersList(Server server) {
        LOGGER.info("Add tot excluded servers list: " + server);
        serversPreference.addToExcludedServersList(server);
    }

    public void removeFromExcludedServerList(Server server) {
        LOGGER.info("Remove tot excluded servers list: " + server);
        serversPreference.removeFromExcludedServerList(server);
    }

    public List<Server> getExcludedServersList() {
        return serversPreference.getExcludedServersList();
    }

    public boolean getSettingFastestServer() {
        return serversPreference.getSettingFastestServer();
    }

    public void putFastestServerSetting(boolean isEnabled) {
        serversPreference.putSettingFastestServer(isEnabled);
    }

    public List<Server> getPossibleServersList() {
        List<Server> excludedServers = getExcludedServersList();
        List<Server> serverList = getCachedServers();

        if (serverList == null) {
            tryUpdateServerListOffline();
            serverList = getCachedServers();
        }

        List<Server> possibleServersList = new ArrayList<>();
        for (Server server : serverList) {
            if (!excludedServers.contains(server)) {
                possibleServersList.add(server);
            }
        }
        if (possibleServersList.size() == 0) {
            possibleServersList.add(getDefaultServer(ServerType.ENTRY));
        }

        return possibleServersList;
    }

    public void addFavouriteServerListener(OnFavouriteServersChangedListener listener) {
        onFavouritesChangedListeners.add(listener);
    }

    public void removeFavouriteServerListener(OnFavouriteServersChangedListener listener) {
        onFavouritesChangedListeners.remove(listener);
    }

    public void addOnServersListUpdatedListener(OnServerListUpdatedListener listener) {
        onServerListUpdatedListeners.add(listener);
    }

    public void removeOnServersListUpdatedListener(OnServerListUpdatedListener listener) {
        onServerListUpdatedListeners.remove(listener);
    }

    private Protocol getCurrentProtocolType() {
        return protocolController.getCurrentProtocol();
    }

    private EnumMap<ServerType, Server> getCurrentServers() {
        Protocol currentProtocol = getCurrentProtocolType();
        return currentServers.get(currentProtocol);
    }

    private List<Server> getSuitableServers(ServersListResponse response) {
        if (getCurrentProtocolType().equals(Protocol.WIREGUARD)) {
            return response.getWireGuardServerList();
        } else {
            return response.getOpenVpnServerList();
        }
    }

    private void notifyFavouriteServerAdded(Server server) {
        for (OnFavouriteServersChangedListener listener : onFavouritesChangedListeners) {
            listener.notifyFavouriteServerAdded(server);
        }
    }

    private void notifyFavouriteServerRemoved(Server server) {
        for (OnFavouriteServersChangedListener listener : onFavouritesChangedListeners) {
            listener.notifyFavouriteServerRemoved(server);
        }
    }
}

File: core/src/main/java/net/ivpn/core/common/prefs/ServersRepository.kt
package net.ivpn.core.common.prefs

/*
IVPN Android app
https://github.com/ivpn/android-app

Created by Oleksandr Mykhailenko.
Copyright (c) 2023 IVPN Limited.

This file is part of the IVPN Android app.

The IVPN Android app is free software: you can redistribute it and/or
modify it under the terms of the GNU General Public License as published by the Free
Software Foundation, either version 3 of the License, or (at your option) any later version.

The IVPN Android app is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
details.

You should have received a copy of the GNU General Public License
along with the IVPN Android app. If not, see <https://www.gnu.org/licenses/>.
*/

import net.ivpn.core.common.Mapper
import net.ivpn.core.common.dagger.ApplicationScope
import net.ivpn.core.rest.HttpClientFactory
import net.ivpn.core.rest.IVPNApi
import net.ivpn.core.rest.RequestListener
import net.ivpn.core.rest.data.ServersListResponse
import net.ivpn.core.rest.data.model.AntiTracker
import net.ivpn.core.rest.data.model.Config
import net.ivpn.core.rest.data.model.Server
import net.ivpn.core.rest.data.model.ServerLocation
import net.ivpn.core.rest.data.model.ServerType
import net.ivpn.core.rest.requests.common.Request
import net.ivpn.core.rest.requests.common.RequestWrapper
import net.ivpn.core.vpn.Protocol
import net.ivpn.core.vpn.ProtocolController
import net.ivpn.core.vpn.controller.VpnBehavior.OnRandomServerSelectionListener
import net.ivpn.core.vpn.model.V2RayPort
import net.ivpn.core.vpn.model.V2RaySettings
import org.slf4j.LoggerFactory
import java.io.Serializable
import java.util.*
import javax.inject.Inject

@ApplicationScope
class ServersRepository @Inject constructor(
        private val settings: Settings,
        private val httpClientFactory: HttpClientFactory,
        private val protocolController: ProtocolController,
        private val serversPreference: ServersPreference,
        private val userPreference: EncryptedUserPreference
): Serializable {

    companion object {
        private val LOGGER = LoggerFactory.getLogger(ServersRepository::class.java)
    }

    private val currentServers = EnumMap<Protocol, EnumMap<ServerType, Server?>>(Protocol::class.java)
    private var onFavouritesChangedListeners: MutableList<OnFavouriteServersChangedListener> = ArrayList()
    private var onServerListUpdatedListeners: MutableList<OnServerListUpdatedListener> = ArrayList()
    private var onServerChangedListeners: List<OnServerChangedListener> = ArrayList()
    private var request: Request<ServersListResponse>? = null

    init {
        currentServers[Protocol.OPENVPN] = EnumMap(ServerType::class.java)
        currentServers[Protocol.WIREGUARD] = EnumMap(ServerType::class.java)
    }

    fun getCurrentServer(serverType: ServerType): Server? {
        var server = serversPreference.getCurrentServer(serverType)
        if (server == null) {
            server = getDefaultServer(serverType)
            setCurrentServer(serverType, server)
        }
        return server
    }

    private fun parseAndSaveV2RaySettings(config: Config?) {
        config?.ports?.v2ray?.let { v2rayConfig ->
            val wireguardPorts = v2rayConfig.wireguard?.map { port ->
                V2RayPort(
                    type = port.protocol,
                    port = port.portNumber
                )
            } ?: emptyList()

            val inboundPort = wireguardPorts.lastOrNull()?.port ?: 0

            val v2raySettings = V2RaySettings(
                id = v2rayConfig.id ?: "",
                outboundIp = "",
                outboundPort = 0,
                inboundIp = "",
                inboundPort = inboundPort,
                dnsName = "",
                wireguard = wireguardPorts
            )

            serversPreference.putV2RaySettings(v2raySettings)
            LOGGER.info("V2Ray settings saved: $v2raySettings")
        }
    }

    fun setCurrentServer(serverType: ServerType, server: Server?) {
        if (server == null) return
        LOGGER.info("Set current server = $server serverType = $serverType")
        getCurrentServers()[serverType] = server
        serversPreference.setCurrentServer(serverType, server)
        val anotherServerType = ServerType.getAnotherType(serverType)
        var serverMH = getCurrentServer(anotherServerType)
        if (!server.canBeUsedAsMultiHopWith(serverMH)) {
            serverMH = getDefaultServer(anotherServerType)
            getCurrentServers()[anotherServerType] = serverMH
            serversPreference.setCurrentServer(anotherServerType, serverMH)
        }
    }

    fun getDefaultServer(serverType: ServerType): Server? {
        val servers = getServers(false)
        if (!servers.isNullOrEmpty()) {
            val anotherServer = serversPreference.getCurrentServer(ServerType.getAnotherType(serverType))
            for (server in servers) {
                if (server.canBeUsedAsMultiHopWith(anotherServer)) {
                    return server
                }
            }
        }
        return null
    }

    fun isServersListExist(): Boolean {
        val servers = serversPreference.serversList
        return servers?.isNotEmpty() ?: false
    }

    fun getServers(isForced: Boolean): List<Server>? {
        var servers = serversPreference.serversList
        if (isForced || servers == null) {
            updateServerList(isForced)
            tryUpdateServerListOffline()
            servers = serversPreference.serversList
        }
        return servers
    }

    val locations: List<ServerLocation>?
        get() {
            var locations = serversPreference.serverLocations
            if (locations == null) {
                updateLocations()
                locations = serversPreference.serverLocations
            }
            return locations
        }

    private fun updateLocations() {
        val servers = serversPreference.serversList
        val locations: MutableList<ServerLocation> = ArrayList()
        if (servers == null) return
        for (server in servers) {
            locations.add(ServerLocation(
                    server.city,
                    server.countryCode,
                    server.latitude,
                    server.longitude
            ))
        }
        if (protocolController.currentProtocol == Protocol.OPENVPN) {
            serversPreference.putOpenVPNLocations(locations)
        } else {
            serversPreference.putWireGuardLocations(locations)
        }
    }

    fun getFavouritesServers(): List<Server> {
        return serversPreference.favouritesServersList
    }

    fun addFavouritesServer(server: Server) {
        LOGGER.info("addFavouritesServer server = $server")
        serversPreference.addFavouriteServer(server)
        notifyFavouriteServerAdded(server)
    }

    fun removeFavouritesServer(server: Server) {
        LOGGER.info("removeFavouritesServer server = $server")
        serversPreference.removeFavouriteServer(server)
        notifyFavouriteServerRemoved(server)
    }

    private fun getCachedServers(): List<Server>? {
        return serversPreference.serversList
    }

    fun getForbiddenServer(serverType: ServerType): Server? {
        val multiHop = settings.isMultiHopEnabled
        return if (!multiHop) null else serversPreference.getCurrentServer(ServerType.getAnotherType(serverType))
    }

    fun updateServerList(isForced: Boolean) {
        LOGGER.info("Updating server list, isForced = $isForced")
        request = Request(settings, httpClientFactory, this, Request.Duration.SHORT, RequestWrapper.IpMode.IPv4)
        request?.start({ obj: IVPNApi -> obj.servers }, object : RequestListener<ServersListResponse> {
            override fun onSuccess(response: ServersListResponse) {
                LOGGER.info("Updating server list, state = SUCCESS_STR")
                LOGGER.info(response.toString())
                response.markServerTypes()
                setServerList(response.openVpnServerList, response.wireGuardServerList)
                updateLocations()
                settings.setIpList(Mapper.stringFromIps(response.config.api.ips))
                settings.setIPv6List(Mapper.stringFromIps(response.config.api.ipv6s))
                settings.wireGuardPorts = response.config.ports.wireguard.filter { it.portNumber > 0 }
                settings.openVpnPorts = response.config.ports.openvpn.filter { it.portNumber > 0 }
                settings.wireGuardPortRanges = response.config.ports.wireguard.filter { it.range != null }
                settings.openVpnPortRanges = response.config.ports.openvpn.filter { it.range != null }
                settings.antiTrackerList = response.config.antiTrackerPlus.list
                if (settings.antiTracker == null) {
                    val defaultDns = AntiTracker()
                    settings.antiTracker = defaultDns.getDefaultList(settings.antiTrackerList, settings, userPreference)
                }
                parseAndSaveV2RaySettings(response.config)
                for (listener in onServerListUpdatedListeners) {
                    listener.onSuccess(getSuitableServers(response), isForced)
                }
            }

            override fun onError(throwable: Throwable) {
                LOGGER.error("Updating server list, state = ERROR")
                for (listener in onServerListUpdatedListeners) {
                    listener.onError(throwable)
                }
            }

            override fun onError(string: String) {
                LOGGER.error("Updating server list, state = ERROR")
                for (listener in onServerListUpdatedListeners) {
                    listener.onError()
                }
            }
        })
    }

    fun fastestServerSelected() {
        serversPreference.putSettingFastestServer(true)
        serversPreference.putSettingRandomServer(false, ServerType.ENTRY)
        for (listener in onServerChangedListeners) {
            listener.onServerChanged()
        }
    }

    fun randomServerSelected(type: ServerType) {
        serversPreference.putSettingRandomServer(true, type)
        if (type == ServerType.ENTRY) {
            serversPreference.putSettingFastestServer(false)
        }
        for (listener in onServerChangedListeners) {
            listener.onServerChanged()
        }
    }

    fun getRandomServerFor(serverType: ServerType, listener: OnRandomServerSelectionListener?) {
        val servers = getServers(false)
        if (!servers.isNullOrEmpty()) {
            val anotherServer = serversPreference.getCurrentServer(ServerType.getAnotherType(serverType))
            val availableServers = servers.filter { it.canBeUsedAsMultiHopWith(anotherServer)}
            val randomServer = availableServers.random()
            setCurrentServer(serverType, randomServer)
            listener?.onRandomServerSelected(randomServer, serverType)
        }
    }

    fun serverSelected(server: Server?, type: ServerType) {
        serversPreference.putSettingFastestServer(false)
        serversPreference.putSettingRandomServer(false, type)
        setCurrentServer(type, server)
        for (listener in onServerChangedListeners) {
            listener.onServerChanged()
        }
    }

    private fun tryUpdateServerListOffline() {
        LOGGER.info("Trying update server list offline from cache...")
        if (getCachedServers() != null) {
            return
        }
        updateServerListOffline()
    }

    fun updateServerListOffline() {
        val response = Mapper.getProtocolServers(ServersLoader.load())
        response?.let{
            it.markServerTypes()
            settings.setIpList(Mapper.stringFromIps(it.config.api.ips))
            settings.setIPv6List(Mapper.stringFromIps(it.config.api.ipv6s))
            settings.wireGuardPorts = response.config.ports.wireguard.filter { it.portNumber > 0 }
            settings.openVpnPorts = response.config.ports.openvpn.filter { it.portNumber > 0 }
            settings.wireGuardPortRanges = response.config.ports.wireguard.filter { it.range != null }
            settings.openVpnPortRanges = response.config.ports.openvpn.filter { it.range != null }
            settings.antiTrackerList = response.config.antiTrackerPlus.list
            if (settings.antiTracker == null) {
                val defaultDns = AntiTracker()
                settings.antiTracker = defaultDns.getDefaultList(settings.antiTrackerList, settings, userPreference)
            }
            parseAndSaveV2RaySettings(response.config)
            setServerList(it.openVpnServerList, it.wireGuardServerList)
        }
    }


    fun loadV2raySettings() {
        val response = Mapper.getProtocolServers(ServersLoader.load())
        response?.let {
            it.markServerTypes()
            settings.setIpList(Mapper.stringFromIps(it.config.api.ips))
            settings.setIPv6List(Mapper.stringFromIps(it.config.api.ipv6s))
            settings.antiTrackerList = it.config.antiTrackerPlus.list

            if (settings.antiTracker == null) {
                val defaultDns = AntiTracker()
                settings.antiTracker = defaultDns.getDefaultList(settings.antiTrackerList, settings, userPreference)
            }

            parseAndSaveV2RaySettings(it.config)
        }
    }


    fun tryUpdateIpList() {
        if (!settings.ipList.isNullOrEmpty()) {
            return
        }
        val response = Mapper.getProtocolServers(ServersLoader.load())
        response?.let {
            settings.wireGuardPorts = response.config.ports.wireguard.filter { it.portNumber > 0 }
            settings.openVpnPorts = response.config.ports.openvpn.filter { it.portNumber > 0 }
            settings.wireGuardPortRanges = response.config.ports.wireguard.filter { it.range != null }
            settings.openVpnPortRanges = response.config.ports.openvpn.filter { it.range != null }
            settings.antiTrackerList = response.config.antiTrackerPlus.list
            if (settings.antiTracker == null) {
                val defaultDns = AntiTracker()
                settings.antiTracker = defaultDns.getDefaultList(settings.antiTrackerList, settings, userPreference)
            }
            parseAndSaveV2RaySettings(response.config)
            settings.setIpList(Mapper.stringFromIps(it.config.api.ips))
            settings.setIPv6List(Mapper.stringFromIps(it.config.api.ipv6s))
        }
    }

    fun tryUpdateServerLocations() {
        LOGGER.info("tryUpdateServerLocations BEFORE")
        if (serversPreference.serverLocations != null) {
            return
        }
        LOGGER.info("tryUpdateServerLocations AFTER")
        val response = Mapper.getProtocolServers(ServersLoader.load())
        response?.let {
            it.markServerTypes()
            val locations: MutableList<ServerLocation> = ArrayList()
            for (server in it.openVpnServerList) {
                locations.add(ServerLocation(
                        server.city,
                        server.countryCode,
                        server.latitude,
                        server.longitude
                ))
            }
            serversPreference.putOpenVPNLocations(locations)
            locations.clear()
            for (server in it.wireGuardServerList) {
                locations.add(ServerLocation(
                        server.city,
                        server.countryCode,
                        server.latitude,
                        server.longitude
                ))
            }
            serversPreference.putWireGuardLocations(locations)
            setServerList(it.openVpnServerList, it.wireGuardServerList)
            updateCurrentServersWithLocation()
            currentServers[Protocol.OPENVPN] = EnumMap(ServerType::class.java)
            currentServers[Protocol.WIREGUARD] = EnumMap(ServerType::class.java)
        }
    }

    private fun updateCurrentServersWithLocation() {
        serversPreference.updateCurrentServersWithLocation()
    }

    fun setLocationList(openVpnLocations: List<ServerLocation>, wireguardLocations: List<ServerLocation>) {
        LOGGER.info("Putting locations, OpenVPN locations list size = " + openVpnLocations.size + " WireGuard = " + wireguardLocations.size)
        serversPreference.putOpenVPNLocations(openVpnLocations)
        serversPreference.putWireGuardLocations(wireguardLocations)
    }

    fun setServerList(openvpnServers: List<Server?>, wireguardServers: List<Server?>) {
        LOGGER.info("Putting servers, OpenVpn servers list size = " + openvpnServers.size + " WireGuard = " + wireguardServers.size)
        serversPreference.putOpenVpnServerList(openvpnServers)
        serversPreference.putWireGuardServerList(wireguardServers)
    }

    fun addToExcludedServersList(server: Server) {
        LOGGER.info("Add tot excluded servers list: $server")
        serversPreference.addToExcludedServersList(server)
    }

    fun removeFromExcludedServerList(server: Server) {
        LOGGER.info("Remove tot excluded servers list: $server")
        serversPreference.removeFromExcludedServerList(server)
    }

    fun getExcludedServersList(): List<Server> {
        return serversPreference.excludedServersList
    }

    fun getSettingFastestServer(): Boolean {
        return serversPreference.settingFastestServer
    }

    fun getSettingRandomServer(serverType: ServerType): Boolean {
        return serversPreference.getSettingRandomServer(serverType)
    }

    fun addOnServersListUpdatedListener(listener: OnServerListUpdatedListener) {
        onServerListUpdatedListeners.add(listener)
    }

    fun removeOnServersListUpdatedListener(listener: OnServerListUpdatedListener) {
        onServerListUpdatedListeners.remove(listener)
    }

    private val currentProtocolType: Protocol
        get() = protocolController.currentProtocol

    private fun getCurrentServers(): EnumMap<ServerType, Server?> {
        val currentProtocol = currentProtocolType
        return currentServers[currentProtocol]!!
    }

    private fun getSuitableServers(response: ServersListResponse): List<Server> {
        return if (currentProtocolType == Protocol.WIREGUARD) {
            response.wireGuardServerList
        } else {
            response.openVpnServerList
        }
    }

    private fun notifyFavouriteServerAdded(server: Server) {
        onFavouritesChangedListeners.forEach { it.notifyFavouriteServerAdded(server) }
    }

    private fun notifyFavouriteServerRemoved(server: Server) {
        onFavouritesChangedListeners.forEach { it.notifyFavouriteServerRemoved(server) }
    }
}
